### DISCUSSION TITLE
Additional status code when no results - 204

### DISCUSSION THREAD

**Discussion by baderdean** [fastapi/fastapi#10035] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def my_func(key: str) -> dict:
  # bla bla bla 
  return {'key': key} if key else None

class JSONorNoneResponse(JSONResponse):
    def render(self, content: any) -> bytes:
        if not content:
            self.status_code = status.HTTP_204_NO_CONTENT
            return None
        return super(JSONorNoneResponse, self).render(content)

# in other part of the code, I manually add the function to the router
# using add_api_route with this specific keyword value
# response_model = (
#                            my_func.__annotations__['return']
#                            | None
#                            ),
```

### Description

Most of the time a function returns a JSON yet sometimes it doesn't find anything good to return and explicitly states it (different from returning null). When a function returns no result, I want it not to return None but return 204 instead. The response type and status code is still usually 200 - JSON for success.

I coined this response_class to handle it and manually modify the response_model also yet I feel that I've hacked in an ugly way. The OpenAPI spec for example doesn't tell about this additional 204 possible Response which annoys me for users especially because I did all of this to make my API more usable to distinguish between void result and NO result.

Is there any elegant pythonic way to do this?

### Operating System

Linux

### FastAPI Version

0.100.0

### Pydantic Version

2.0.3

### Python Version

3.11.2

**Comment by n8sty** [comment]
Check out this portion of the documentation https://fastapi.tiangolo.com/advanced/additional-responses/ for some guidance on extending the possible OpenAPI responses for a route operation.

**Reply by baderdean** [reply]
Thanks, it almost worked, almost! I've added:
```
responses= {204: {'model': None}}
```
as an add_api_route parameter, yet know OpenAPI specs explains that the return could be a "string"

Yet the OpenAPI spec is clear about it, it should be NO value at all:
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#response-object-examples

So is it a bug? Or is there any way to circumvent this?

**Reply by baderdean** [reply]
And it doesn't address the issue about the custom object Response that I created and my hack to stop validation error by manually adding None as a return type in the response_model parameter
